### ISSUE TITLE
Equivalent of sarama.NewReferenceHashPartitioner

### ISSUE THREAD

**Issue by fracasula** [segmentio/kafka-go#905]
Hello, I was reading in your documentation and I quote:

> If you're switching from Sarama and need/want to use the same algorithm for message partitioning, you can use the kafka.Hash balancer. kafka.Hash routes messages to the same partitions that Sarama's default partitioner would route to.

Which is pretty straightforward. The default sarama partitioner seems to be [sarama.NewHashPartitioner](https://github.com/Shopify/sarama/blob/main/config.go#L477).

However, due to a bug, they created `NewReferenceHashPartitioner`:
> // NewReferenceHashPartitioner is like NewHashPartitioner except that it handles absolute values
> // in the same way as the reference Java implementation. NewHashPartitioner was supposed to do
> // that but it had a mistake and now there are people depending on both behaviors. This will
> // all go away on the next major version bump.

https://github.com/Shopify/sarama/blob/main/partitioner.go#L170

Is there an equivalent of `NewReferenceHashPartitioner` in this repo?

Also, given the above it looks like `kafka.Hash` = `sarama.NewHashPartitioner` and the sarama one is bugged.

Shouldn't we get rid of the same bug here and change [this](https://github.com/segmentio/kafka-go/blob/4da3b721ca38db775a5024089cdc4ba14f84e698/balancer.go#L162) to [this](https://github.com/Shopify/sarama/blob/a37f45dcee6b1e95240c17a4c60c4a960d50eaeb/partitioner.go#L201)?

**Comment by jahidmixmaster** [comment]
@fracasula I think kafka-go already implemented this. Have a look [here](https://github.com/segmentio/kafka-go/blob/4da3b721ca38db775a5024089cdc4ba14f84e698/balancer.go#L245)

**Comment by fracasula** [comment]
> @fracasula I think kafka-go already implemented this. Have a look [here](https://github.com/segmentio/kafka-go/blob/4da3b721ca38db775a5024089cdc4ba14f84e698/balancer.go#L245)

I'll look into it thanks @jahidmixmaster :+1: 
Also I was using a round robin which is not correct, a random partitioner should be used like in the one you linked (i.e. murmur2).

**Comment by fracasula** [comment]
> > @fracasula I think kafka-go already implemented this. Have a look [here](https://github.com/segmentio/kafka-go/blob/4da3b721ca38db775a5024089cdc4ba14f84e698/balancer.go#L245)
> 
> I'll look into it thanks @jahidmixmaster +1 Also I was using a round robin which is not correct, a random partitioner should be used like in the one you linked (i.e. murmur2).

@jahidmixmaster I built a test and I verified that the Murmur2 does not produce the same results as I expected:

```go
func TestReferenceHashBalancer(t *testing.T) {
	testCases := map[string]struct {
		Key        []byte
		Hasher     hash.Hash32
		Partitions []int
		Partition  int
	}{
		"partition-0": {
			Key:        []byte("blah"),
			Partitions: []int{0, 1},
			Partition:  0,
		},
		"partition-1": {
			Key:        []byte("blah"),
			Partitions: []int{0, 1, 2},
			Partition:  1,
		},
		"partition-2": {
			Key:        []byte("castle"),
			Partitions: []int{0, 1, 2},
			Partition:  2,
		},
		"hash code with MSB set": {
			Key:        []byte("20"),
			Partitions: []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
			Partition:  15,
		},
	}

	saramaPartitioner := sarama.NewReferenceHashPartitioner("")

	for label, test := range testCases {
		t.Run(label, func(t *testing.T) {
			msg := Message{Key: test.Key}
			h := ReferenceHash{Hasher: test.Hasher}
			//h := Murmur2Balancer{Consistent: true}
			segmentPartition := h.Balance(msg, test.Partitions...)
			saramaPartition, err := saramaPartitioner.Partition(
				&sarama.ProducerMessage{Key: sarama.ByteEncoder(test.Key)},
				int32(len(test.Partitions)),
			)
			if err != nil {
				t.Errorf("got error: %v", err)
			}
			if saramaPartition != int32(segmentPartition) {
				t.Errorf("expected %v; got %v", saramaPartition, segmentPartition)
			}
		})
	}
}
```

Feel try to run the test. Then run it again by uncommenting `h := Murmur2Balancer{Consistent: true}` (try also with `Consistent=false`, it won't matter).

The reason why is that the Sarama partitioner uses `fnv1aPool.Get().(hash.Hash32)` as the hasher VS the one you proposed that is using `murmur2`.

**Comment by rhansen2** [comment]
Closed via #906.
